### PR TITLE
Download enterprise repository via GitHub token

### LINKED ISSUE #2: git authentication
cloning right now (https://github.com/grafana/grafana-build/blob/main/containers/clone.go) implicitly uses `$HOME/.ssh/id_rsa` to attempt to clone `grafana-enterprise`.

We should use another authentication method.

Some options are:

1. Add the `github cli` as a requirement, and use the result of the `gh auth token` command.
2. Using an alpine image with the Github CLI installed and the `os.UserConfigDir()` function, mount the `gh` folder into a container at `/root/.config/gh`, then use the GitHub CLI to clone the repository.
3. Prompt the user for an https username / password
4. Require a GH_TOKEN to be set.

I would prefer 2, as it matches the pattern used with the google-cloud-sdk container. But it's annoying because GitHub doesn't and won't distribute an official docker image, so we have to download their release. https://github.com/cli/cli/issues/2027

Requirements:
* When running a command in this repository that requires git authentication, I should not have to provide any more information than I would with a normal `git clone` operation.

### PR DESCRIPTION
**Pull request by zerok**
Resolves #2

### CODE REVIEW

File: cmd/github.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/urfave/cli/v2"
+)
+
+// lookupGitHubToken will try to find a GitHub access token that can then be used for various API calls but also cloning of private repositories.
+func lookupGitHubToken(cctx *cli.Context) (string, error) {
```
Comment (kminehart) [inline cmd/github.go:15]:
```suggestion
func lookupGitHubToken(c *cli.Context) (string, error) {
```

File: cmd/github.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	log.Print("Looking for a GitHub token")
+	token := cctx.String("github-token")
```
Comment (kminehart) [inline cmd/github.go:17]:
```suggestion
	token := c.String("github-token")
```

File: cmd/github.go (hunk 1 above)
Comment (kminehart) [reply cmd/github.go:15]: Change to be consistent with other cli.Context arguments in the project

File: cmd/github.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	if token != "" {
+		log.Print("Using GitHub token provided via flag")
+		return token, nil
+	}
+	// First try: Check if it's in the environment. This can override everything!
+	token = os.Getenv("GITHUB_TOKEN")
+	if token != "" {
+		log.Print("Using GitHub token provided via environment variable")
+		return token, nil
+	}
```
Comment (kminehart) [inline cmd/github.go:27]: I like this and I was thinking about this earlier, how we should prioritize this over mounting the 'gh' directory or trying to use the 'gh' command.

File: cmd/github.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		err = cmd.Run()
+		if err != nil {
+			log.Printf("Querying gh for an access token failed: %s", errData.String())
+			return "", fmt.Errorf("lookup in gh failed: %w", err)
+		}
+		log.Print("Using GitHub token provided via gh")
+		return strings.TrimSpace(data.String()), nil
+	}
+
+	return "", nil
```
Comment (kminehart) [inline cmd/github.go:46]:
```suggestion
	ghPath, err := exec.LookPath("gh")
	if err != nil {
		return "", nil
	}
	
	cmd := exec.CommandContext(cctx.Context, ghPath, "auth", "token")
	cmd.Stdout = &data
	cmd.Stderr = &errData
	
	if err := cmd.Run(); err != nil {
		log.Printf("Querying gh for an access token failed: %s", errData.String())
		return "", fmt.Errorf("lookup in gh failed: %w", err)
	}
	
	log.Print("Using GitHub token provided via gh")
	return strings.TrimSpace(data.String()), nil
```

I just inversed the conditions here to be more consistent with the rest of the project.

Should this condition return an error?

```
	ghPath, err := exec.LookPath("gh")
	if err != nil {
		return "", nil
	}
```

I'm fine with returning an error here saying something like, "Expected a `--github-token` flag, `GITHUB_TOKEN` environment variable, or a configured GitHub CLI (`$XDG_CONFIG_HOME/gh` or `%appdata%/gh`)"

**Review by kminehart — CHANGES_REQUESTED** [review]
Looks good to me, the logic makes sense. I requested a few changes for consistency.

File: cmd/github.go (hunk 3 above)
Comment (zerok) [reply cmd/github.go:27]: Among others. It's also the environment variable that would be picked up by `gh` if it's set and pretty much all other tooling that somehow interacts with GitHub.

File: cmd/github.go (hunk 4 above)
Comment (zerok) [reply cmd/github.go:46]: Personally, I think it would be nicer to have that as fall-through response: `return "", fmt.Errorf("expected ...")`

**Review by kminehart — APPROVED** [review]

**Pull request merged**